lib/node_modules/@stdlib/blas/base/dgemm/lib/base.js

```diff
@@ -0,0 +1,259 @@
+* @returns {Float64Array} `C`
+*
+* @example
+* var Float64Array = require( '@stdlib/array/float64' );
+*
+* var A = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
+* var B = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0 ] );
+* var C = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
+*
+* dgemm( 'row-major', 'none', 'none', 2, 3, 3, 5.0, A, 3, 0, B, 3, 0, 6.0, C, 3, 0 );
+* // C => <Float64Array>[ 156.0, 192.0, 228.0, 354.0, 435.0, 516.0 ]
+*/
+function dgemm( order, transa, transb, m, n, k, alpha, A, LDA, offsetA, B, LDB, offsetB, beta, C, LDC, offsetC ) { // eslint-disable-line max-len, max-params
+	var nota;
+	var notb;
+	var tmp;
+	var i;
+	var j;
+	var l;
+
+	nota = transa === 'none';
+	notb = transb === 'none';
+
+	if ( m === 0 || n === 0 || ( ( ( alpha === 0.0 || k === 0 ) ) && beta === 1.0 ) ) { // eslint-disable-line max-len
+		return C;
+	}
+
+	if ( alpha === 0.0 ) {
+		if ( beta === 0.0 ) {
+			if ( order === 'column-major' ) {
+				for ( j = 0; j < n; j++ ) {
+					for ( i = 0; i < m; i++ ) {
+						C[ offsetC + ( j * LDC ) + i ] = 0.0;
+					}
+				}
+				return C;
+			}
+			// order === 'row-major'
+			for ( j = 0; j < n; j++ ) {
+				for ( i = 0; i < m; i++ ) {
+					C[ offsetC + ( i * LDC ) + j ] = 0.0;
+				}
+			}
+			return C;
+		}
+		// beta !== 0.0
+		if ( order === 'column-major' ) {
+			for ( j = 0; j < n; j++ ) {
+				for ( i = 0; i < m; i++ ) {
+					C[ offsetC + ( j * LDC ) + i ] *= beta;
+				}
+			}
+			return C;
+		}
+		// order === 'row-major'
+		for ( j = 0; j < n; j++ ) {
+			for ( i = 0; i < m; i++ ) {
+				C[ offsetC + ( i * LDC ) + j ] *= beta;
+			}
+		}
+		return C;
+	}
+	if ( notb ) {
+		if ( nota ) {
+			// C := alpha * A * B + beta * C
+			if ( order === 'column-major' ) {
+				for ( j = 0; j < n; j++ ) {
+					if ( beta === 0.0 ) {
+						for ( i = 0; i < m; i++ ) { // eslint-disable-line max-depth
+							C[ offsetC + ( j * LDC ) + i ] = 0.0;
+						}
+					} else if ( beta !== 1.0 ) {
+						for ( i = 0; i < m; i++ ) { // eslint-disable-line max-depth
+							C[ offsetC + ( j * LDC ) + i ] *= beta;
+						}
+					}
+					for ( l = 0; l < k; l++ ) {
+						tmp = alpha * B[ offsetB + ( j * LDB ) + l ];
+						for ( i = 0; i < m; i++ ) {
+							C[ offsetC + ( j * LDC ) + i ] += tmp * A[ offsetA + ( l * LDA ) + i ]; // eslint-disable-line max-len
+						}
+					}
+				}
+				return C;
+			}
+			// order === 'row-major'
+			for ( j = 0; j < n; j++ ) {
+				if ( beta === 0.0 ) {
+					for ( i = 0; i < m; i++ ) {
+						C[ offsetC + ( i * LDC ) + j ] = 0.0;
+					}
+				} else if ( beta !== 1.0 ) {
+					for ( i = 0; i < m; i++ ) {
+						C[ offsetC + ( i * LDC ) + j ] *= beta;
+					}
+				}
+				for ( l = 0; l < k; l++ ) {
+					tmp = alpha * B[ offsetB + ( l * LDB ) + j ];
+					for ( i = 0; i < m; i++ ) {
+						C[ offsetC + ( i * LDC ) + j ] += tmp * A[ offsetA + ( i * LDA ) + l ]; // eslint-disable-line max-len
+					}
+				}
+			}
+			return C;
+		}
```

Comment:
When you see this code, the only difference is accessing and writing matrices `A`, `B` and `C`. How about a utility function with API:

```javascript
function stdlib_blas_base_matrix_read( order, A, LDA, offsetA, colA, rowA )
```

Where `colA` and `rowA` are the column and row index of element to be fetched and based on this we can fetch value.

Similar API to write

```javascript
function stdlib_blas_base_matrix_write( order, A, LDA, offsetA, colA, rowA, value )
```

We can figure out variable and function naming if idea seems fine.

cc @kgryte

Reply:
I am not sure about cache optimisation in these cases. My basic intuition states that on generalising it for both orders we won't be able to optimise the `row-major` case ( reading and writing via cache ). I'll be happy to read any article / blog in these directions.

Reply:
This won't work due to inefficient traversals and you'd have nested conditionals, thus making compiler optimizations more difficult. The primary way you address this is through loop interchange. See base ndarray utilities (e.g., `ndarray/base/assign`) for an example.